File: aether/python/selenium_core/_webdriver.py
"""The ergonomic Python WebDriver surface — a thin layer over the Aether core.

Carries NO protocol logic: the command catalog, route table, path templating,
By normalization and W3C error decode all live in the engine (``core/``). Every
command here is one ``_native.execute`` call plus JSON marshalling. This is the
Python-idiomatic face (``driver.get(url)``, ``driver.find_element(By.CSS, ...)``)
that a user touches.
"""

from __future__ import annotations

import json
from typing import Any

from . import _native

# The W3C element-reference key. A findElement result is
# {"element-6066-11e4-a52e-4f735466cecf": "<id>"}.
_W3C_ELEMENT_KEY = "element-6066-11e4-a52e-4f735466cecf"


class By:
    """Locator strategies. Values match the engine's ``by_locator`` strategy
    strings; id/name/class name are rewritten to CSS in the engine."""

    ID = "id"
    NAME = "name"
    CSS_SELECTOR = "css selector"
    CLASS_NAME = "className"
    TAG_NAME = "tag name"
    LINK_TEXT = "link text"
    PARTIAL_LINK_TEXT = "partial link text"
    XPATH = "xpath"


class WebDriverError(Exception):
    """Base for all remote-end errors, matching the W3C error taxonomy."""

    def __init__(self, msg: str = "", code: int = 0):
        super().__init__(msg)
        self.w3c_code = code


class NoSuchElementError(WebDriverError):
    pass


class StaleElementReferenceError(WebDriverError):
    pass


class ElementClickInterceptedError(WebDriverError):
    pass


class ElementNotInteractableError(WebDriverError):
    pass


class InvalidSelectorError(WebDriverError):
    pass


class NoSuchWindowError(WebDriverError):
    pass


class NoSuchFrameError(WebDriverError):
    pass


class TimeoutError_(WebDriverError):
    pass


class JavascriptError(WebDriverError):
    pass


class SessionNotCreatedError(WebDriverError):
    pass


class UnknownCommandError(WebDriverError):
    pass


# The engine's integer error codes (selenium_core.error_code) -> exception type.
# Codes not listed map to the base WebDriverError.
_CODE_TO_EXC = {
    3: ElementClickInterceptedError,
    4: ElementNotInteractableError,
    11: InvalidSelectorError,
    13: JavascriptError,
    17: NoSuchElementError,
    18: NoSuchFrameError,
    20: NoSuchWindowError,
    21: TimeoutError_,
    22: SessionNotCreatedError,
    23: StaleElementReferenceError,
    24: TimeoutError_,
    28: UnknownCommandError,
}


def _raise_for(code: int, message: str) -> None:
    exc = _CODE_TO_EXC.get(code, WebDriverError)
    raise exc(message, code)


class WebElement:
    """A remote element handle. Methods issue element-scoped commands via the
    engine, passing this element's id as the ``:id`` path parameter."""

    def __init__(self, driver: "WebDriver", element_id: str):
        self._driver = driver
        self._id = element_id

    @property
    def id(self) -> str:
        return self._id

    def _exec(self, command: str, params: dict | None = None) -> Any:
        p = dict(params or {})
        p["id"] = self._id
        return self._driver._execute(command, p)

    def click(self) -> None:
        self._exec("clickElement")

    def clear(self) -> None:
        self._exec("clearElement")

    def send_keys(self, text: str) -> None:
        # W3C expects {"text": full, "value": [chars...]} — send both for
        # broad driver compatibility.
        self._exec("sendKeysToElement", {"text": text, "value": list(text)})

    @property
    def text(self) -> str:
        return self._exec("getElementText")

    @property
    def tag_name(self) -> str:
        return self._exec("getElementTagName")

    def get_attribute(self, name: str) -> Any:
        # DOM attribute (W3C native endpoint). The atom-based getAttribute
        # (property-or-attribute) is a binding/engine follow-up.
        return self._exec("getDomAttribute", {"name": name})

    def get_property(self, name: str) -> Any:
        return self._exec("getElementProperty", {"name": name})

    def value_of_css_property(self, name: str) -> Any:
        return self._exec("getElementValueOfCssProperty", {"propertyName": name})

    def is_enabled(self) -> bool:
        return bool(self._exec("isElementEnabled"))

    def is_selected(self) -> bool:
        return bool(self._exec("isElementSelected"))

    @property
    def rect(self) -> dict:
        return self._exec("getElementRect")

    def screenshot_as_base64(self) -> str:
        return self._exec("takeElementScreenshot")

    def find_element(self, by: str, value: str) -> "WebElement":
        loc = _decode_by(by, value)
        loc["id"] = self._id
        result = self._driver._execute("findChildElement", loc)
        return WebElement(self._driver, result[_W3C_ELEMENT_KEY])

    def find_elements(self, by: str, value: str) -> list["WebElement"]:
        loc = _decode_by(by, value)
        loc["id"] = self._id
        result = self._driver._execute("findChildElements", loc)
        return [WebElement(self._driver, e[_W3C_ELEMENT_KEY]) for e in result]

    def __eq__(self, other):
        return isinstance(other, WebElement) and other._id == self._id

    def __repr__(self):
        return f"<WebElement id={self._id!r}>"


def _decode_by(by: str, value: str) -> dict:
    """Ask the engine for the {"using","value"} locator (shares the ONE By
    normalization + CSS-escape path with every other binding)."""
    raw = _native.take_string(_native.by_locator(_native.encode(by), _native.encode(value)))
    return json.loads(raw)


class WebDriver:
    """A WebDriver session. Construct with a remote-end URL and capabilities,
    or use :func:`Chrome` / :func:`Remote` helpers."""

    def __init__(self, command_executor: str, capabilities: dict | None = None):
        self._handle = _native.open(_native.encode(command_executor))
        caps = capabilities or {"browserName": "chrome"}
        # W3C newSession envelope: {"capabilities": {"alwaysMatch": {...}}}
        payload = {"capabilities": {"alwaysMatch": caps}}
        self._execute("newSession", payload)

    # ---- the FFI seam ----

    def _execute(self, command: str, params: dict | None = None) -> Any:
        params_json = json.dumps(params or {})
        rc = _native.execute(
            self._handle, _native.encode(command), _native.encode(params_json)
        )
        if rc != 0:
            code = _native.last_error_code(self._handle)
            message = _native.take_string(_native.last_error(self._handle))
            if rc == -1 and code == 0:
                # transport-level failure
                raise WebDriverError(message or "transport failure", -1)
            _raise_for(code, message)
        raw = _native.take_string(_native.last_value(self._handle))
        if raw == "":
            return None
        return json.loads(raw)

    # ---- navigation ----

    def get(self, url: str) -> None:
        self._execute("get", {"url": url})

    @property
    def current_url(self) -> str:
        return self._execute("getCurrentUrl")

    @property
    def title(self) -> str:
        return self._execute("getTitle")

    @property
    def page_source(self) -> str:
        return self._execute("getPageSource")

    def back(self) -> None:
        self._execute("goBack")

    def forward(self) -> None:
        self._execute("goForward")

    def refresh(self) -> None:
        self._execute("refresh")

    # ---- elements ----

    def find_element(self, by: str, value: str) -> WebElement:
        result = self._execute("findElement", _decode_by(by, value))
        return WebElement(self, result[_W3C_ELEMENT_KEY])

    def find_elements(self, by: str, value: str) -> list[WebElement]:
        result = self._execute("findElements", _decode_by(by, value))
        return [WebElement(self, e[_W3C_ELEMENT_KEY]) for e in result]

    # ---- script ----

    def execute_script(self, script: str, *args) -> Any:
        return self._execute("executeScript", {"script": script, "args": list(args)})

    def execute_async_script(self, script: str, *args) -> Any:
        return self._execute("executeAsyncScript", {"script": script, "args": list(args)})

    # ---- windows ----

    @property
    def window_handles(self) -> list[str]:
        return self._execute("getWindowHandles")

    @property
    def current_window_handle(self) -> str:
        return self._execute("getCurrentWindowHandle")

    def switch_to_window(self, handle: str) -> None:
        self._execute("switchToWindow", {"handle": handle})

    def maximize_window(self) -> None:
        self._execute("maximizeWindow")

    def minimize_window(self) -> None:
        self._execute("minimizeWindow")

    def fullscreen_window(self) -> None:
        self._execute("fullscreenWindow")

    def get_window_rect(self) -> dict:
        return self._execute("getWindowRect")

    def set_window_rect(self, x=None, y=None, width=None, height=None) -> dict:
        rect = {}
        if x is not None:
            rect["x"] = x
        if y is not None:
            rect["y"] = y
        if width is not None:
            rect["width"] = width
        if height is not None:
            rect["height"] = height
        return self._execute("setWindowRect", rect)

    # ---- cookies ----

    def add_cookie(self, cookie: dict) -> None:
        self._execute("addCookie", {"cookie": cookie})

    def get_cookies(self) -> list[dict]:
        return self._execute("getCookies")

    def get_cookie(self, name: str) -> dict:
        return self._execute("getCookie", {"name": name})

    def delete_cookie(self, name: str) -> None:
        self._execute("deleteCookie", {"name": name})

    def delete_all_cookies(self) -> None:
        self._execute("deleteAllCookies")

    # ---- alerts ----

    def accept_alert(self) -> None:
        self._execute("acceptAlert")

    def dismiss_alert(self) -> None:
        self._execute("dismissAlert")

    @property
    def alert_text(self) -> str:
        return self._execute("getAlertText")

    def send_alert_text(self, text: str) -> None:
        self._execute("setAlertValue", {"text": text})

    # ---- timeouts ----

    def set_page_load_timeout(self, ms: int) -> None:
        self._execute("setTimeout", {"pageLoad": ms})

    def set_script_timeout(self, ms: int) -> None:
        self._execute("setTimeout", {"script": ms})

    def implicitly_wait(self, ms: int) -> None:
        self._execute("setTimeout", {"implicit": ms})

    # ---- screenshots ----

    def get_screenshot_as_base64(self) -> str:
        return self._execute("screenshot")

    # ---- lifecycle ----

    @property
    def session_id(self) -> str:
        return _native.take_string(_native.session_id(self._handle))

    def quit(self) -> None:
        try:
            self._execute("quit")
        finally:
            self.close_handle()

    def close(self) -> None:
        """Close the current window (not the whole session)."""
        self._execute("close")

    def close_handle(self) -> None:
        if self._handle:
            _native.close(self._handle)
            self._handle = None

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.quit()


def Chrome(command_executor: str = "http://127.0.0.1:9515", options: dict | None = None) -> WebDriver:
    """Convenience: a Chrome session against a running chromedriver.

    ``options`` is a raw capabilities dict merged under ``browserName: chrome``
    (e.g. ``{"goog:chromeOptions": {"args": ["--headless=new"]}}``)."""
    caps = {"browserName": "chrome"}
    if options:
        caps.update(options)
    return WebDriver(command_executor, caps)


def Remote(command_executor: str, capabilities: dict) -> WebDriver:
    """A session against an arbitrary remote end / Grid hub with explicit caps."""
    return WebDriver(command_executor, capabilities)

File: ruby/lib/selenium_core/webdriver.rb
# frozen_string_literal: true

require 'json'
require_relative 'native'

module SeleniumCore
  # Locator strategies. Values match the engine's by_locator strategy strings;
  # ID/NAME/CLASS_NAME are rewritten to CSS in the engine.
  module By
    ID                = 'id'
    NAME              = 'name'
    CSS_SELECTOR      = 'css selector'
    CLASS_NAME        = 'className'
    TAG_NAME          = 'tag name'
    LINK_TEXT         = 'link text'
    PARTIAL_LINK_TEXT = 'partial link text'
    XPATH             = 'xpath'
  end

  # Base for all remote-end errors, carrying the engine's stable W3C error code
  # (0 = success, -1 = transport failure).
  class WebDriverError < StandardError
    attr_reader :code

    def initialize(message = '', code = 0)
      super(message)
      @code = code
    end
  end

  class NoSuchElementError < WebDriverError; end
  class StaleElementReferenceError < WebDriverError; end
  class ElementClickInterceptedError < WebDriverError; end
  class ElementNotInteractableError < WebDriverError; end
  class InvalidSelectorError < WebDriverError; end
  class TimeoutError < WebDriverError; end
  class JavascriptError < WebDriverError; end
  class UnknownCommandError < WebDriverError; end

  # Engine integer error codes -> exception class (see core error_code()).
  CODE_TO_EXC = {
    3  => ElementClickInterceptedError,
    4  => ElementNotInteractableError,
    11 => InvalidSelectorError,
    13 => JavascriptError,
    17 => NoSuchElementError,
    21 => TimeoutError,
    23 => StaleElementReferenceError,
    24 => TimeoutError,
    28 => UnknownCommandError
  }.freeze

  W3C_ELEMENT_KEY = 'element-6066-11e4-a52e-4f735466cecf'

  # ---- pure engine helpers (no session needed) ----
  module_function

  # "METHOD PATH" route for a command name, or "" if unknown.
  def route(command)
    Native.take_string(Native.call(:route, command))
  end

  # Map a W3C error string to its stable integer code (0 = success).
  def error_code(w3c_error)
    Native.call(:error_code, w3c_error)
  end

  # The W3C {"using","value"} locator JSON for a (by, value) pair.
  def locator(by, value)
    Native.take_string(Native.call(:by_locator, by, value))
  end

  def decode_by(by, value)
    JSON.parse(locator(by, value))
  end

  # A remote element handle. Methods issue element-scoped commands, passing this
  # element's id as the :id path parameter.
  class WebElement
    attr_reader :id

    def initialize(driver, id)
      @driver = driver
      @id = id
    end

    def click
      exec('clickElement')
    end

    def clear
      exec('clearElement')
    end

    def send_keys(text)
      exec('sendKeysToElement', 'text' => text, 'value' => text.chars)
    end

    def text
      exec('getElementText')
    end

    def tag_name
      exec('getElementTagName')
    end

    # Whether the element is shown (the isDisplayed atom, run in-page by the
    # engine — the visibility algorithm, not a naive style check).
    def displayed?
      !!@driver.send(:atom_bool, :is_displayed, @id)
    end

    # The classic getAttribute(name): property-or-attribute (boolean attrs,
    # live properties like value/checked), via the shared engine atom. Use
    # #dom_attribute for the raw W3C DOM attribute.
    def attribute(name)
      @driver.send(:atom_get_attribute, @id, name)
    end

    # The literal DOM attribute (W3C getDomAttribute), no property fallback.
    def dom_attribute(name)
      exec('getDomAttribute', 'name' => name)
    end

    def property(name)
      exec('getElementProperty', 'name' => name)
    end

    def rect
      exec('getElementRect')
    end

    def enabled?
      !!exec('isElementEnabled')
    end

    def selected?
      !!exec('isElementSelected')
    end

    def find_element(by, value)
      params = SeleniumCore.decode_by(by, value)
      params['id'] = @id
      result = @driver.send(:execute, 'findChildElement', params)
      WebElement.new(@driver, result.fetch(W3C_ELEMENT_KEY))
    end

    def ==(other)
      other.is_a?(WebElement) && other.id == @id
    end

    private

    def exec(command, params = {})
      params = params.merge('id' => @id)
      @driver.send(:execute, command, params)
    end
  end

  # A WebDriver session over the shared engine.
  class WebDriver
    # Start a Chrome session against a running chromedriver (or Grid).
    #   options: a raw capabilities Hash merged under browserName: chrome.
    def self.chrome(command_executor = 'http://127.0.0.1:9515', options: nil)
      caps = { 'browserName' => 'chrome' }
      caps.merge!(options) if options
      new(command_executor, caps)
    end

    # Convenience: headless-Chrome launch args baked in.
    def self.headless_chrome(command_executor = 'http://127.0.0.1:9515')
      chrome(command_executor, options: {
               'goog:chromeOptions' => {
                 'args' => ['--headless=new', '--no-sandbox', '--disable-gpu', '--disable-dev-shm-usage']
               }
             })
    end

    def initialize(command_executor, capabilities)
      @handle = Native.call(:open, command_executor)
      raise WebDriverError.new('failed to open session handle', -1) if @handle.nil? || @handle.null?

      # Request a BiDi channel so #bidi is available on demand; the WebSocket
      # itself opens lazily (a classic script never opens it).
      caps = capabilities.merge('webSocketUrl' => true)
      result = execute('newSession', 'capabilities' => { 'alwaysMatch' => caps })
      @ws_url = (result.is_a?(Hash) ? result.dig('capabilities', 'webSocketUrl') : nil) || ''
      @bidi = nil
    end

    # ---- navigation ----
    def get(url)
      execute('get', 'url' => url)
      nil
    end

    def current_url = execute('getCurrentUrl')
    def title        = execute('getTitle')
    def page_source  = execute('getPageSource')
    def back  = (execute('goBack'); nil)
    def forward = (execute('goForward'); nil)
    def refresh = (execute('refresh'); nil)

    # ---- elements ----
    def find_element(by, value)
      result = execute('findElement', SeleniumCore.decode_by(by, value))
      WebElement.new(self, result.fetch(W3C_ELEMENT_KEY))
    end

    def find_elements(by, value)
      result = execute('findElements', SeleniumCore.decode_by(by, value))
      result.map { |e| WebElement.new(self, e.fetch(W3C_ELEMENT_KEY)) }
    end

    # ---- script ----
    def execute_script(script, *args)
      execute('executeScript', 'script' => script, 'args' => args)
    end

    # ---- atom-backed commands (run a shared JS atom in-page via the engine) ----

    # Relative locators: elements matching +base_css+ filtered by spatial
    # relation to anchors, nearest first. Each filter is a Hash
    # {kind: "above"|"below"|"left"|"right"|"near", sel: "<css>"} (+near+ also
    # accepts +dist+). Returns an Array of WebElement.
    def find_relative(base_css, *filters)
      rc = Native.call(:find_relative, @handle, base_css, JSON.generate(filters))
      result = atom_result(rc) || []
      result.map { |ref| WebElement.new(self, ref.fetch(W3C_ELEMENT_KEY)) }
    end

    # ---- windows ----
    def window_handles = execute('getWindowHandles')
    def current_window_handle = execute('getCurrentWindowHandle')
    def maximize_window = (execute('maximizeWindow'); nil)
    def minimize_window = (execute('minimizeWindow'); nil)

    # ---- cookies ----
    def add_cookie(cookie) = (execute('addCookie', 'cookie' => cookie); nil)
    def cookies = execute('getCookies')
    def cookie(name) = execute('getCookie', 'name' => name)
    def delete_cookie(name) = (execute('deleteCookie', 'name' => name); nil)
    def delete_all_cookies = (execute('deleteAllCookies'); nil)

    # ---- window rect ----
    def window_rect = execute('getWindowRect')
    def set_window_rect(rect) = execute('setWindowRect', rect)

    # ---- W3C actions ----
    def perform_actions(actions) = (execute('actions', 'actions' => actions); nil)
    def clear_actions = (execute('clearActions'); nil)

    # ---- screenshots ----
    def screenshot_base64 = execute('screenshot')

    # ---- lifecycle ----
    def session_id
      Native.take_string(Native.call(:session_id, @handle))
    end

    # ---- WebDriver-BiDi ----

    # True if this session negotiated a webSocketUrl (BiDi usable).
    def bidi_available? = !@ws_url.empty?

    # The event-driven BiDi surface for this session, opened lazily over the
    # negotiated webSocketUrl. Raises if the remote end granted no BiDi URL.
    #
    #   driver.bidi.subscribe(BidiEvent::LOG_ENTRY_ADDED)
    #   driver.get(url)
    #   ev = driver.bidi.next_event(BidiEvent::LOG_ENTRY_ADDED, timeout_ms: 5000)
    def bidi
      @bidi ||= begin
        raise WebDriverError.new('BiDi not available: no webSocketUrl negotiated', 0) if @ws_url.empty?

        handle = Native.call(:bidi_open, @ws_url)
        raise WebDriverError.new('BiDi channel failed to open', -1) if handle.nil? || handle.null?

        BiDi.new(handle)
      end
    end

    def quit
      @bidi&.close
      @bidi = nil
      execute('quit')
    ensure
      close_handle
    end

    private

    def close_handle
      return unless @handle && !@handle.null?

      Native.call(:close, @handle)
      @handle = nil
    end

    # The FFI seam: one command by name with a Hash of JSON params. Returns the
    # decoded `value` payload, or raises a typed WebDriverError.
    def execute(command, params = {})
      rc = Native.call(:execute, @handle, command, JSON.generate(params))
      if rc != 0
        code = Native.call(:last_error_code, @handle)
        message = Native.take_string(Native.call(:last_error, @handle))
        if rc == -1 && code.zero?
          raise WebDriverError.new(message.empty? ? 'transport failure' : message, -1)
        end

        exc = CODE_TO_EXC.fetch(code, WebDriverError)
        raise exc.new(message, code)
      end
      raw = Native.take_string(Native.call(:last_value, @handle))
      return nil if raw.empty?

      JSON.parse(raw)
    end

    # Drain last_value after an atom call, raising the same typed WebDriverError
    # that #execute uses on a non-zero return code.
    def atom_result(rc)
      if rc != 0
        code = Native.call(:last_error_code, @handle)
        message = Native.take_string(Native.call(:last_error, @handle))
        if rc == -1 && code.zero?
          raise WebDriverError.new(message.empty? ? 'transport failure' : message, -1)
        end

        exc = CODE_TO_EXC.fetch(code, WebDriverError)
        raise exc.new(message, code)
      end
      raw = Native.take_string(Native.call(:last_value, @handle))
      return nil if raw.empty?

      JSON.parse(raw)
    end

    # An atom verb taking (handle, element_id) whose last_value is a JSON boolean.
    def atom_bool(verb, element_id)
      !!atom_result(Native.call(verb, @handle, element_id))
    end

    # The getAttribute atom: (handle, element_id, name) -> JSON string|null.
    def atom_get_attribute(element_id, name)
      atom_result(Native.call(:get_attribute, @handle, element_id, name))
    end
  end

  # The common WebDriver-BiDi event names (W3C spec). Pass to
  # +driver.bidi.subscribe(...)+ and match in +next_event+.
  module BidiEvent
    LOG_ENTRY_ADDED     = 'log.entryAdded'
    CONTEXT_CREATED     = 'browsingContext.contextCreated'
    CONTEXT_DESTROYED   = 'browsingContext.contextDestroyed'
    NAVIGATION_STARTED  = 'browsingContext.navigationStarted'
    DOM_CONTENT_LOADED  = 'browsingContext.domContentLoaded'
    LOAD                = 'browsingContext.load'
    DOWNLOAD_WILL_BEGIN = 'browsingContext.downloadWillBegin'
    BEFORE_REQUEST_SENT = 'network.beforeRequestSent'
    AUTH_REQUIRED = 'network.authRequired'
    RESPONSE_STARTED    = 'network.responseStarted'
    RESPONSE_COMPLETED  = 'network.responseCompleted'
    FETCH_ERROR         = 'network.fetchError'
    REALM_CREATED       = 'script.realmCreated'
    REALM_DESTROYED     = 'script.realmDestroyed'
    MESSAGE             = 'script.message'
  end

  # The event-driven BiDi channel for a session (over the demux C ABI).
  # Commands and events multiplex over one WebSocket via the engine's shape-C
  # demux (single reader -> id-keyed replies + bounded event queue), so replies
  # stay correlated while events stream. Command ids are supplied automatically.
  class BiDi
    def initialize(handle)
      @handle = handle
      @next_id = 1
    end

    # session.subscribe to one or more event names; wait for the ack. Returns the
    # ack payload Hash. Matching events then arrive on the queue (drain via
    # #next_event).
    def subscribe(*events, timeout_ms: 10_000)
      raw = Native.take_string(Native.call(:bidi_subscribe, @handle, next_id, events.join(','), timeout_ms))
      raw.empty? ? {} : JSON.parse(raw)
    end

    def unsubscribe(*events, timeout_ms: 10_000)
      raw = Native.take_string(Native.call(:bidi_unsubscribe, @handle, next_id, events.join(','), timeout_ms))
      raw.empty? ? {} : JSON.parse(raw)
    end

    # Block until an event whose +method+ matches arrives, or timeout. Returns the
    # event Hash, or nil on timeout/close. (Subscribe first.)
    def next_event(method, timeout_ms: 5_000)
      raw = Native.take_string(Native.call(:bidi_wait_event, @handle, method, timeout_ms))
      raw.empty? ? nil : JSON.parse(raw)
    end

    # Issue any BiDi command and return its reply payload Hash. Reaches BiDi
    # methods with no dedicated wrapper (script.evaluate, network.*, ...).
    def command(method, params = {}, timeout_ms: 10_000)
      cid = next_id
      raise WebDriverError.new("BiDi send failed: #{method}", -1) if
        Native.call(:bidi_send, @handle, cid, method, JSON.generate(params)) != 0

      waited = 0
      step = 50
      while waited < timeout_ms
        reply = Native.take_string(Native.call(:bidi_poll_reply, @handle, cid))
        return JSON.parse(reply) unless reply.empty?
        break if Native.call(:bidi_pump, @handle, step) < 0

        waited += step
      end
      raise TimeoutError.new("BiDi command timed out: #{method}", 0)
    end

    # ---- typed convenience commands ----

    # browsingContext.getTree — the browsing contexts (each with a "context" id).
    def get_tree(timeout_ms: 10_000)
      raw = Native.take_string(Native.call(:bidi_get_tree, @handle, next_id, timeout_ms))
      raw.empty? ? {} : JSON.parse(raw)
    end

    # The top-level browsing context id (the anchor for evaluate/navigate), or nil.
    def top_context(timeout_ms: 10_000)
      contexts = get_tree(timeout_ms: timeout_ms).dig('result', 'contexts') || []
      contexts.empty? ? nil : contexts.first['context']
    end

    # script.evaluate an expression in a context's realm, awaiting a returned
    # promise. Returns the reply Hash; dig('result', 'result') is the BiDi-typed
    # value (e.g. {"type" => "number", "value" => 42}). BiDi's richer alternative
    # to execute_script — real realms, promise-awaiting, structured value types.
    def evaluate(expr, context: nil, timeout_ms: 30_000)
      ctx = context || top_context(timeout_ms: timeout_ms)
      raise WebDriverError.new('no browsing context for script.evaluate', 0) unless ctx

      raw = Native.take_string(Native.call(:bidi_script_evaluate, @handle, next_id, expr, ctx, timeout_ms))
      raw.empty? ? {} : JSON.parse(raw)
    end

    # script.evaluate, returning just the unwrapped value (the .value of the
    # BiDi-typed result), or nil if it wasn't a simple value.
    def evaluate_value(expr, context: nil, timeout_ms: 30_000)
      evaluate(expr, context: context, timeout_ms: timeout_ms).dig('result', 'result', 'value')
    end

    # browsingContext.navigate a context to url (wait: complete). Returns the reply.
    def navigate(url, context: nil, timeout_ms: 30_000)
      ctx = context || top_context(timeout_ms: timeout_ms)
      raise WebDriverError.new('no browsing context for navigate', 0) unless ctx

      raw = Native.take_string(Native.call(:bidi_navigate, @handle, next_id, ctx, url, timeout_ms))
      raw.empty? ? {} : JSON.parse(raw)
    end

    # ---- network interception (observe / release / block requests) ----

    # network.addIntercept for a URL pattern (a full parseable URL as a "string"
    # pattern; empty intercepts all) at the given comma-separated phases.
    # Subscribe to the matching network.* event first if you want the
    # paused-request events. Returns the intercept id String, or nil.
    def add_intercept(phases: 'beforeRequestSent', url_pattern: '', timeout_ms: 10_000)
      raw = Native.take_string(
        Native.call(:bidi_network_add_intercept, @handle, next_id, phases, url_pattern, timeout_ms)
      )
      reply = raw.empty? ? {} : JSON.parse(raw)
      reply.dig('result', 'intercept')
    end

    def remove_intercept(intercept_id, timeout_ms: 10_000)
      raw = Native.take_string(
        Native.call(:bidi_network_remove_intercept, @handle, next_id, intercept_id, timeout_ms)
      )
      raw.empty? ? {} : JSON.parse(raw)
    end

    # Let a paused (intercepted) request proceed unchanged. request_id comes from
    # a network event's +params.request.request+ (see .event_request_id).
    def continue_request(request_id, timeout_ms: 10_000)
      raw = Native.take_string(
        Native.call(:bidi_network_continue_request, @handle, next_id, request_id, timeout_ms)
      )
      raw.empty? ? {} : JSON.parse(raw)
    end

    # Block a paused request (the ad/tracker-blocking case).
    def fail_request(request_id, timeout_ms: 10_000)
      raw = Native.take_string(
        Native.call(:bidi_network_fail_request, @handle, next_id, request_id, timeout_ms)
      )
      raw.empty? ? {} : JSON.parse(raw)
    end

    # Fulfill a paused request with a mock response — the request never hits the
    # network. request_id comes from a network event's +params.request.request+
    # (see .event_request_id). The engine adds Access-Control-Allow-Origin:* to
    # the mock so a cross-origin page can read the body. Returns the reply Hash.
    def provide_response(request_id, status: 200, content_type: '', body: '', timeout_ms: 10_000)
      raw = Native.take_string(
        Native.call(:bidi_network_provide_response, @handle, next_id, request_id, status, content_type, body,
                    timeout_ms)
      )
      raw.empty? ? {} : JSON.parse(raw)
    end

    # Answer an HTTP auth challenge (a paused authRequired request) with
    # credentials — automates basic/digest auth that classic WebDriver can't
    # handle in headless. Returns the reply Hash.
    def continue_with_auth(request_id, username, password, timeout_ms: 10_000)
      raw = Native.take_string(
        Native.call(:bidi_network_continue_with_auth, @handle, next_id, request_id, username, password, timeout_ms)
      )
      raw.empty? ? {} : JSON.parse(raw)
    end

    # Set the session HTTP cache behavior: "bypass" to disable it (so every
    # request hits the network / an intercept), "default" to restore it.
    # Returns the reply Hash.
    def set_cache_behavior(behavior = 'bypass', timeout_ms: 10_000)
      raw = Native.take_string(
        Native.call(:bidi_network_set_cache_behavior, @handle, next_id, behavior, timeout_ms)
      )
      raw.empty? ? {} : JSON.parse(raw)
    end

    # The network.request id out of a network.beforeRequestSent (or other
    # network) event: +params.request.request+.
    def self.event_request_id(event)
      event.dig('params', 'request', 'request')
    end

    # How many events the bounded queue dropped since the last call (then resets).
    def lost_events = Native.call(:bidi_lost_events, @handle)

    def close
      return unless @handle && !@handle.null?

      Native.call(:bidi_close, @handle)
      @handle = nil
    end

    private

    def next_id
      id = @next_id
      @next_id += 1
      id
    end
  end
end
